### PR TITLE
Add initial Archlinux and LDAP support

### PR DESCRIPTION
**Pull request by bastelfreak**

### CODE REVIEW

File: manifests/params.pp [hunk 1]
```diff
@@ -35,6 +35,12 @@
       $servicename      = 'dhcpd'
       $package_provider = undef
     }
+    'Archlinux': {
+      $dhcp_dir         = '/etc'
```
Comment (bastelfreak) [inline manifests/params.pp]: there is no /etc/dhcp on Archlinux.

**Comment by bastelfreak** [comment]
This module is a bit itchy at all corners, but in general, the Arch support is now working. I will have a look at the spec tests.

File: manifests/params.pp (hunk 1 above)
Comment (nibalizer) [inline manifests/params.pp]: Bastel reviews his own code I guess?

Comment (bastelfreak) [inline manifests/params.pp]: He does!
I'm not sure how I want to workaround the broken defaults from upstream. The prefered way seems to be to throw everything in one file, called /etc/dhcpd.conf. The default for other distributions is to put the content into seperate files in /etc/dhcp/...

Comment (bastelfreak) [inline manifests/params.pp]: So my prefered way in general is to stay with upstream. But in that case that would be very itchy. I would like to create a /etc/dhcp/ directory and throwevery file in it and simply adjust the default service file.

File: manifests/params.pp [hunk 2]
```diff
@@ -39,7 +39,7 @@
       $dhcp_dir         = '/etc'
       $packagename      = 'dhcp'
       $servicename      = ['dhcpd4.service', 'dhcpd6.service']
-      $package_provider = 'systemd'
```
Comment (igalic) [inline manifests/params.pp]: lol

File: manifests/params.pp (hunk 1 above)
Comment (bastelfreak) [inline manifests/params.pp]: I reconsidered this and would like to keep it as it is.

**Comment by igalic** [comment]
how is this still failing after a modulesync??

**Comment by bastelfreak** [comment]
I just synced the Gemfile, not everything. I will put this thing on hold and work on the modulesync PR.

File: spec/classes/dhcp_spec.rb [hunk 3]
```diff
@@ -156,21 +156,45 @@
...
+      it do
+        content = catalogue.resource('concat::fragment', 'dhcp-conf-ldap').send(:parameters)[:content]
+        expected_lines = [
+          'ldap-port 389;',
+          'ldap-server "localhost";',
+          'ldap-username "cn=root, dc=example, dc=com";',
+          'ldap-password "passw0rd";',
+          'ldap-base-dn "dc=example, dc=com";',
+          'ldap-method dynamic;',
+          'ldap-debug-file "/var/log/dhcp-ldap-startup.log"'
```
Comment (alexjfisher) [inline spec/classes/dhcp_spec.rb]: missing a ';' at the end of this line

File: manifests/init.pp [hunk 4]
```diff
@@ -9,7 +9,7 @@
   $dhcp_conf_ntp        = 'INTERNAL_TEMPLATE',
   $dhcp_conf_pxe        = 'INTERNAL_TEMPLATE',
   $dhcp_conf_extra      = 'INTERNAL_TEMPLATE',
-  $dhcp_conf_fragments  = {},
+  $dhcp_conf_fragments = {},
```
Comment (alexjfisher) [inline manifests/init.pp]: nitpick. What happened here?

File: manifests/init.pp [hunk 5]
```diff
@@ -207,6 +215,37 @@
...
+    if ($ldap_server == '') {
+      fail('you must set $ldap_server')
+    }
+    if ($ldap_username == '') {
+      fail('you must set $ldap_username')
+    }
+    if ($ldap_base_dn == '') {
+      fail('you must set $ldap_username')
+    }
+    if ($ldap_debug_file == '') {
```
Comment (alexjfisher) [inline manifests/init.pp]: validate_absolute_path ?

File: templates/dhcpd.conf-ldap.erb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+# BEGIN LDAP
+ldap-port <%= @ldap_port %>;
+ldap-server "<%= @ldap_server %>";
+ldap-username "<%= @ldap_username %>";
+ldap-password "<%= @ldap_password %>";
+ldap-base-dn "<%= @ldap_base_dn %>";
+ldap-method <%= @ldap_method %>;
+ldap-debug-file "<%= @ldap_debug_file %>";
```
Comment (alexjfisher) [inline templates/dhcpd.conf-ldap.erb]: Perhaps some way of turning this off?
`$ldap_debug_enable` or somesuch?

**Comment by alexjfisher** [comment]
Maybe split into 2 PRs (or rebase to two commits); one to add ArchLinux support and one for adding ldap functionality.

File: manifests/init.pp (hunk 4 above)
Comment (bastelfreak) [inline manifests/init.pp]: rebase fuckup. fixed in cad5d4afab0dbf1f4fe728deb5031ae8690d47ee

File: manifests/init.pp (hunk 5 above)
Comment (bastelfreak) [inline manifests/init.pp]: as mentioned in the commit message, validate_absolute_path needs a path to a directory, not to a filename so I removed it after I first added it.

File: templates/dhcpd.conf-ldap.erb (hunk 6 above)
Comment (bastelfreak) [inline templates/dhcpd.conf-ldap.erb]: what about allowing '' as value and only write the line if the var isn't empty?

File: manifests/init.pp (hunk 5 above)
Comment (alexjfisher) [inline manifests/init.pp]: really?  Pretty sure I've used/seen it used with filename paths loads of times.
eg
https://github.com/puppetlabs/puppetlabs-ntp/blob/master/manifests/init.pp#L62

File: templates/dhcpd.conf-ldap.erb (hunk 6 above)
Comment (alexjfisher) [inline templates/dhcpd.conf-ldap.erb]: Works for me.

**Comment by bastelfreak** [comment]
@alexjfisher I think handling this in one PR is fine, the changes are all seperate commits and they should all be single logical changes. Can we merge it as is?

**Comment by alexjfisher** [comment]
24 commits seems a bit excessive, but sure, the code looks good :)

**Comment by jyaworski** [comment]
@bastelfreak is it still WIP? The title says it is.

File: templates/dhcpd.conf-ldap.erb (hunk 6 above)
Comment (bastelfreak) [inline templates/dhcpd.conf-ldap.erb]: done

**Comment by bastelfreak** [comment]
@jyaworski sorry, fixed

**Pull request merged**